### ISSUE TITLE
Filter NSFW cards for (unpinned) top sites and highlights

### ISSUE THREAD

**Issue by edwinwongsf** [mozilla/activity-stream#1195]
Eng bug for: https://trello.com/c/iDJ1IY5e

Need Engineering design and architecture. We should consider consuming a NSFW blacklist on the server.

**Comment by pdehaan** [comment]
Do we know if NSFW sites/pages need to self-identify w/ headers or meta tags, or is a server list the best/only solution?

**Comment by piatra** [comment]
This seems like a good option https://developers.google.com/safe-browsing/

**Comment by tspurway** [comment]
@ncloudioj, is there other public blacklists that you are aware of for NSFW content?

**Comment by ncloudioj** [comment]
Not to my knowledge. Usually it needs some support from the 3rd party services, such as Adsafe.

**Comment by pdehaan** [comment]
> ## What are the Safe Browsing APIs?

> The Safe Browsing APIs (v4) let your client applications check URLs against Google's constantly updated lists of unsafe web resources. Examples of unsafe web resources are social engineering sites (phishing and deceptive sites) and sites that host malware or unwanted software. Any URL found on a Safe Browsing list is considered unsafe.

Not sure if those will cover NSFW sites, or just potentially malicious sites (although the two may overlap).

**Comment by edmoz** [comment]
Fx already consumes google's safe browsing API which is primarily for malware.  We're looking for NSFW blacklists.  @nchapman has implemented this before and should be familiar with such services

**Comment by pdehaan** [comment]
Yeah, looking at the safebrowsing API, it doesn't look very useful, or detect NSFW sites at all:
https://gist.github.com/pdehaan/56700418071713b46a806cd5431ba95a

Embed.ly's `/extract` API does return this field though, but I think we've discussed several times and decided it wasn't worthwhile to add to our metadata proxy:

> `safe` &mdash; Safe is an attribute that tells you if the url is on a phishing or malware list. Embedly uses Google's Safe Browsing API to obtain a list of malicious urls. By using this API through our service, you agree to its terms of service.

> &mdash; via http://docs.embed.ly/docs/extract#response-attributes

**Comment by edmoz** [comment]
there could be leads to such services here: http://dansguardian.org/?page=whatisdg

**Comment by jaredkerim** [comment]
Backlogging this pending further discussion.

**Comment by edmoz** [comment]
aaron is going to start a draft to define this work and UX mtg will review.

**Comment by pdehaan** [comment]
@nchapman mentioned he may have a list, but I also quickly threw together https://github.com/pdehaan/alexa-top-sites the other day for some testing. At a minimum we could scrape the first X pages of [Alexa "Top/Adult"](http://www.alexa.com/topsites/category/Top/Adult) pages.

**Comment by pdehaan** [comment]
Another fun part of this problem is that we'd potentially need to block other _image hosting sites_.

I also quickly (and poorly) wrote [**pdehaan/fetch-subreddit**](https://github.com/pdehaan/fetch-subreddit) which takes a subreddit name (or array of names) and returns the links for the first page of content. Not very interesting, but once you start scanning the 10 most popular subreddits by subscriber count (via http://redditlist.com/), you can start doing some `.reduce()` fun to get [this list](https://gist.github.com/pdehaan/a3ba1d0aca739f14da3c89e752d9002f) of content hosts.

**TL;DR:** common content hosting sites [for Reddit] include:

```js
{
  "i.imgur.com": 111,
  "gfycat.com": 72,
  "imgur.com": 35,
  "www.reddit.com": 4,
  "www.youtube.com": 4,
  "youtu.be": 3,
  "m.imgur.com": 2,
  ...
}
```

Not to mention Tumblr, etc.

**Comment by pdehaan** [comment]
Re: Tumblr, https://github.com/mozilla/page-metadata-service/issues/97 "Ignore NSFW content from Tumblr"

> It's only a drop in the bucket, but it looks like Tumblr at least returns an `X-Tumblr-Content-Rating: nsfw` header which we can probably filter on after getting the `fetch()` response.

**Comment by pdehaan** [comment]
In the news today: https://techcrunch.com/2016/09/30/yahoo-open-sources-its-porn-detecting-neural-network/

Although, probably extreme overkill, and may only apply if a URL links directly to an image, so we can't infer meta tags or run the scraper against the URL since there is no DOM.

**Comment by aaronrbenson** [comment]
Combine the following two lists (checking for dupes) into a block list that would stop any Top Site or Highlight from being automatically generated. 

http://www.alexa.com/topsites/category;0/Top/Adult

https://github.com/mozilla/heatmap/blob/master/pornfilter/domain_blocklist.txt

**Comment by Mardak** [comment]
There's a list of domains already shipping in Firefox available: `chrome://browser/content/newtab/newTab.inadjacent.json`

https://dxr.mozilla.org/mozilla-central/search?q=inadjacent.json
https://bugzilla.mozilla.org/show_bug.cgi?id=1159884

**Comment by pdehaan** [comment]
@Mardak Neat. Is there an easy way to do a base64 decode or whatever to make the newTab.inadjacent.json readable?

Mine (in Nightly) looks like:

```json
{
    "domains": [
        "rp5slFCxq/e7hYhXJCd0vQ==",
        "2rEimAJDNX5g8HPZehOrGg==",
        "nvLEpj6ZZF3LWH3wUB6lKg==",
        "9Cqd4Lm3VvXuJxz79Bbqyg==",
```

**Comment by Mardak** [comment]
They were generated as 1-way hashes using `md5` to allow comparison with a known domain, i.e., a given link to show/hide: https://dxr.mozilla.org/mozilla-central/source/browser/modules/DirectoryLinksProvider.jsm#967-991

I believe the source domains are here: https://github.com/matthewruttley/contentfilter/blob/048e1410da2408d4e79ef0ea8001691606fe9af4/sites.json

**Comment by tspurway** [comment]
The plan here is to use the Alexa top 500 'adult' sub-category.  Grab the list of domains from Maria.  We should obfuscate the actual domains by hashing them.  When we do the query, we need to remove any Highlight that comes from these sites.

**Comment by Mardak** [comment]
Content Services had an inadjacent.json file:
https://hg.mozilla.org/mozilla-central/log/default/browser/base/content/newtab/newTab.inadjacent.json

It has hashed domains that we can has the alexa list to verify that they're included or add them. It was noted that the alexa list doesn't include some sites that should be categorized as adult.

**Comment by Mardak** [comment]
The desired functionality is to filter out everything except pinned top sites

**Comment by nt1m** [comment]
Apparently some people rely on this feature: https://www.reddit.com/r/firefox/comments/7e09ya/new_tab_top_sites_hides_porn/ :)
